[AutoDiff] Use ASTMangler for generated AD associated functions.

dan-zheng opened the pull request:
Use ASTMangler for generated AD associated functions:
JVPs, VJPs, differentials, pullbacks.

This centralizes the ad-hoc mangling logic scattered throughout the codebase.

Todos:
- [TF-680](https://bugs.swift.org/browse/TF-680): Mangle `@differentiable` attribute requirements.
- [TF-685](https://bugs.swift.org/browse/TF-685): Use ASTMangler for AD-related thunks.
- [TF-686](https://bugs.swift.org/browse/TF-686): Make Demangler/Remangler work with AD-generated functions.

dan-zheng committed:
[AutoDiff] Use ASTMangler for generated AD associated functions.

Use ASTMangler for generated AD associated functions:
JVPs, VJPs, differentials, pullbacks.

This centralizes ad-hoc mangling logic scattered throughout the codebase.

Todos:
- TF-680: Mangle `@differentiable` attribute requirements.
- TF-685: Use ASTMangler for AD-related thunks.
- TF-686: Make Demangler/Remangler work with AD-generated functions.

dan-zheng commented:
cc @bartchr808 
This PR will cause conflicts with https://github.com/apple/swift/pull/26057, I'm happy to help resolve them together.

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented on include/swift/AST/ASTMangler.h:
```diff
@@ -151,7 +151,17 @@ class ASTMangler : public Mangler {
+  // - Associated function kind.
+  // - Parameter/result indices.
+  // - Flag representing whether the function is a differential/pullback.
+  std::string mangleAutoDiffAssociatedFunctionHelper(
```
Associated function kind is orthogonal to linear maps. I’d suggest adding to separate mangling helpers, one for associates functions and one for linear maps. I think you should also add a LinearMapKind enum to distinguish between differential and pullback.

dan-zheng replied:
Added `AutoDiffLinearMapKind` in https://github.com/apple/swift/pull/26389/commits/221ddb0d76529c6f043b98af2e3734bdb5682be9. Let me know if you prefer the name `LinearMapKind`.

Added `ASTMangler::mangleAutoDiffLinearMapHelper` in https://github.com/apple/swift/pull/26389/commits/c5b2018a91fe40a34c44d4e7a164fc96f19d34b9.

bartchr808 commented on lib/AST/ASTMangler.cpp:
```diff
@@ -369,6 +369,37 @@ std::string ASTMangler::mangleReabstractionThunkHelper(
+
+  Buffer << "AD__";
+  Buffer << name;
+  Buffer << '_';
```
Could these three lines be combined into: `Buffer << "AD__" << name << '_';`

dan-zheng replied:
Done in https://github.com/apple/swift/pull/26389/commits/c5b2018a91fe40a34c44d4e7a164fc96f19d34b9.

dan-zheng committed:
Address review feedback.

- Create separate `ASTMangler::mangleAutoDiffLinearMapHelper` function.

dan-zheng committed:
Add todo references for TF-685.

dan-zheng committed:
Add `AutoDiffLinearMapKind` enum.

Add `AutoDiffLinearMapKind` enum, use it in
`ASTMangler::mangleAutoDiffLinearMapHelper`.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented on include/swift/AST/ASTMangler.h:
```diff
@@ -167,7 +167,7 @@ class ASTMangler : public Mangler {
   // - Associated function kind.
```
Linear map kind.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/26389/commits/724aa0cf67c5902d95b8ef8b362ece8cba96b664.

dan-zheng committed:
Address review feedback.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow
